File: Data Structures & Algorithms/two-integer-sum/submission-8.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:

        visited = {}

        for i in range(len(nums)):
            result = target - nums[i]

            if result in visited:
                return [visited.get(result), i]
            else:
                visited[nums[i]] = i
        return [-1, -1]

            
        # [3, 4, 5, 6] target = 7

        # 7 - 3 -> 3 -> 0
        # 4 - 7 -> 

File: Data Structures & Algorithms/combination-target-sum/submission-8.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(result, target, new ArrayList<>(), 0, nums);
        return result;

    }

    public void backtrack(List<List<Integer>> result, int target, List<Integer> current, int i, int[] nums){
        if(target == 0){
            result.add(new ArrayList<>(current));
            return;
        }

        if(target < 0 || i >= nums.length){
            return;
        }

        current.add(nums[i]);
        backtrack(result, target - nums[i], current, i, nums);
        current.remove(current.size() - 1);
        backtrack(result, target, current, i + 1, nums);
    }
}
